src/indexing/enclave_lfs_access.h

```diff
@@ -0,0 +1,282 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ccf/indexing/lfs_interface.h"
+#include "crypto/entropy.h"
+#include "crypto/hash.h"
+#include "crypto/symmetric_key.h"
+#include "ds/hex.h"
+#include "ds/messaging.h"
+
+#include <optional>
+#include <set>
+#include <unordered_map>
+
+// Uncomment to disable encryption and obfuscation, writing cache content
+// directly unencrypted to host disk
+// #define PLAINTEXT_CACHE
+
+namespace ccf::indexing
+{
+  static inline std::vector<uint8_t> get_iv(const LFSKey& key)
+  {
+    auto h = crypto::SHA256((const uint8_t*)key.data(), key.size());
+    h.resize(crypto::GCM_SIZE_IV);
+    return h;
+  }
+
+  static inline LFSEncryptedContents encrypt(
+    crypto::KeyAesGcm& encryption_key,
+    const LFSKey& key,
+    LFSContents&& contents)
+  {
+    crypto::GcmCipher gcm(contents.size());
+    auto iv = get_iv(key);
+    gcm.hdr.set_iv(iv.data(), iv.size());
+
+    encryption_key.encrypt(
+      gcm.hdr.get_iv(), contents, nullb, gcm.cipher.data(), gcm.hdr.tag);
+
+#ifdef PLAINTEXT_CACHE
+    gcm.cipher = contents;
+#endif
+
+    return gcm.serialise();
+  }
+
+  static inline bool verify_and_decrypt(
+    crypto::KeyAesGcm& encryption_key,
+    const LFSKey& key,
+    LFSEncryptedContents&& encrypted,
+    std::vector<uint8_t>& plaintext)
+  {
+    crypto::GcmCipher gcm;
+    gcm.deserialise(encrypted);
+
+    const CBuffer given_iv = gcm.hdr.get_iv();
+    const auto expected_iv = get_iv(key);
+    if (
+      given_iv.n != expected_iv.size() ||
+      (memcmp(given_iv.p, expected_iv.data(), given_iv.n) != 0))
+    {
+      LOG_TRACE_FMT(
+        "IV mismatch for {}: {:02x} != {:02x}",
+        key,
+        fmt::join(given_iv.p, given_iv.p + given_iv.n, " "),
+        fmt::join(get_iv(key), " "));
+      return false;
+    }
+
+#ifdef PLAINTEXT_CACHE
+    plaintext = gcm.cipher;
+    const auto success = true;
+#else
+    plaintext.resize(gcm.cipher.size());
+    const auto success = encryption_key.decrypt(
+      gcm.hdr.get_iv(), gcm.hdr.tag, gcm.cipher, nullb, plaintext.data());
+#endif
+
+    if (success)
+    {
+      return true;
+    }
+    else
+    {
+      LOG_TRACE_FMT("Decryption failed for {}", key);
+      return false;
+    }
+  }
+
+  class EnclaveLFSAccess : public AbstractLFSAccess
+  {
+  protected:
+    using PendingResult = std::weak_ptr<FetchResult>;
+
+    std::unordered_map<LFSKey, PendingResult> pending;
```

Comment:
To minimise memory copies and memory use by this API, it stores results in a caller-managed `FetchResult`. So to drop a request, the caller simply drops their handle (and then our weak ptr lock fails). And there's no 2-step check-for-result + read-contents process, instead the caller has the result and contents together (modulo some locking, which I haven't implemented yet).